numpy/_core/src/multiarray/convert_datatype.c

```diff
@@ -125,43 +103,104 @@ PyArray_GetCastingImpl(PyArray_DTypeMeta *from, PyArray_DTypeMeta *to)
             PyArray_VectorUnaryFunc *castfunc = PyArray_GetCastFunc(
                     from->singleton, to->type_num);
             if (castfunc == NULL) {
-                PyErr_Clear();
-                /* Remember that this cast is not possible */
-                if (PyDict_SetItem(NPY_DT_SLOTS(from)->castingimpls,
-                            (PyObject *) to, Py_None) < 0) {
-                    return NULL;
-                }
```

Comment:
`None` gets inserted into to castingimpls later on before exiting the critical section so there's no need to do it here